### PR TITLE
Exclude Go vendor files from rat configuration

### PR DESCRIPTION
**Pull request by jbonofre**
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [X] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [X] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [X] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [X] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

**Comment by jbonofre** [comment]
retest this please

**Comment by herohde** [comment]
Thanks JB! I see I excluded it from gradle's RAT check, but forgot about maven:

https://github.com/apache/beam/commit/6a72d79978c9e186e5b8abd81ac701901aa05a52

Not sure if we need to use the wider regex used there for "vendor" and ".gogradle". The Go container boot code used for java and python might produce such files, for example.

**Comment by jbonofre** [comment]
I aligned the rat configuration in Maven and Gradle.

@kennknowles rat uses the `defaultExcludes` but it doesn't read `.gitignore`. That's why we have the "`.gitignore`" exclude in the main `pom.xml`.

**Comment by herohde** [comment]
LGTM

Thanks for fixing this issue, JB. Sorry for missing it.

**Comment by jbonofre** [comment]
No problem ! Team work ! Thanks for the review ! Just waiting green light from Jenkins before merging.

**Comment by jbonofre** [comment]
Maven build needs #4259 to work (fix javadoc on `fn-execution`) to pass.

**Comment by kennknowles** [comment]
What I mean is this:

```
$ mvn -Drat.skip=false apache-rat:check -X
...
[INFO] --- apache-rat-plugin:0.12:check (default-cli) @ beam-parent ---
...
[DEBUG] Adding plexus default exclusions...
[DEBUG] Adding SCM default exclusions...
[DEBUG] Adding exclusions often needed by Maven projects...
[DEBUG] Adding exclusions often needed by projects developed in Eclipse...
[DEBUG] Adding exclusions often needed by projects developed in IDEA...
[INFO] Will parse SCM ignores for exclusions...
[INFO] Parsing exclusions from /usr/local/google/home/klk/GitHub/apache/beam/.gitignore
[DEBUG] Added 
[DEBUG] Added .gradle/
[DEBUG] Added .gogradle/
[DEBUG] Added build/
[DEBUG] Added vendor/
[DEBUG] Added 
[DEBUG] Added bin/
[DEBUG] Added dependency-reduced-pom.xml
[DEBUG] Added target/
[DEBUG] Added 
[DEBUG] Added sdks/java/maven-archetypes/examples/src/main/resources/archetype-resources/src/
[DEBUG] Added sdks/java/maven-archetypes/examples-java8/src/main/resources/archetype-resources/src/
[DEBUG] Added 
...
```

In #4027 I added a bunch of exclusions to the `pom.xml` but they turned out to be redundant and the problem was something else IIRC.

There is a subtlety that I only just sorted out fully. The patterns in a `.gitignore` are resolved relative to the `.gitignore` but _also_ directories inherit the `.gitignore` from parent directories.

So when `.gitignore` contains `vendor/` this means that when `git` is working in `sdks/go` it will inherit the `vendor/` ignore pattern, and it will work. But when the working directory is the root, `sdks/go/vendor` is not ignored.

I think it is probably best to just have full patterns in `.gitignore`, too. It also supports the same `**/vendor/` syntax.

**Comment by jbonofre** [comment]
@kennknowles good point ! It means that we can update .gitignore with the correct patterns and then delegate to rat (both in Maven and Gradle). Let me try this. Thanks !

**Comment by lukecwik** [comment]
Note that the gradle implementation is using the Apache Ant Rat task and not the Maven plugin and when I investigated the Ant plugin it seemed to have less features then the Maven version.

**Comment by jbonofre** [comment]
What should we do about this PR ? Should I remove the rat configuration to relay on `.gitignore` ?

**Comment by lukecwik** [comment]
I think using `.gitignore` would be best. Note that the gradle build uses the Ant RAT task which doesn't look at `.gitignore` and still needs to have things listed twice, once for the Rat plugin configurtion and once inside `.gitignore`.

**Comment by jbonofre** [comment]
OK, I update the Maven RAT plugin configuration then. Thanks !

**Comment by jbonofre** [comment]
I updated `.gitignore` to use `**/*` syntax. Then, Maven `apache-rat-plugin` uses `.gitignore` plus some specific file we track but not distribute (as in gradle).

File: build.gradle [hunk 1]
```diff
@@ -187,10 +187,10 @@ rat {
...
-    "**/.gogradle/**",
-    "**/build/**",
-    "**/vendor/**",
+    "**/.gradle/**/*",
+    "**/.gogradle/**/*",
+    "**/build/**/*",
+    "**/vendor/**/*",
     "**/.gradletasknamecache",
 
     // .gitignore: Ignore files generated by the Maven build process
```
Comment (lukecwik) [inline build.gradle]: remove the duplicate `**/build/**/*` directory below like you did in the .gitignore

Also, `sdks/python/**/*.egg` is missing from the list below

**Pull request merged**